src/riscv/lib/src/jit/builder/instr_map.rs

```diff
@@ -0,0 +1,144 @@
+        }
+    }
+
+    /// Insert a new address into the map.
+    pub(super) fn insert(&mut self, address: Address) {
+        self.addresses.push(address);
+    }
+
+    /// Retrieve the address at the provided instr ID.
+    pub(super) fn get(&self, instr_id: InstrId) -> Address {
+        self.addresses[instr_id.0]
+    }
+
+    /// Get the instr ID of the instruction at the given address.
+    pub(super) fn get_instr_id(&self, address: Address) -> Option<InstrId> {
+        self.addresses.binary_search(&address).ok().map(InstrId)
+    }
+}
+
+/// A holder of the instructions in a sequence with quick lookup via instr ID or address.
+#[derive(Debug, Clone)]
+pub(super) struct InstrMap<V> {
+    instructions: Vec<V>,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(super) struct InstrId(usize);
+
+impl<V> InstrMap<V> {
+    /// Retrieve the instruction at the provided instr ID.
+    pub(super) fn get(&self, instr_id: InstrId) -> &V {
+        &self
+            .instructions
+            .get(instr_id.0)
+            .expect("instrID out of bounds.")
+    }
+
+    /// Retrieve a mutable reference to the instruction at the provided instr ID.
+    pub(super) fn get_mut(&mut self, instr_id: InstrId) -> &mut V {
+        self.instructions
+            .get_mut(instr_id.0)
+            .expect("instrID out of bounds.")
+    }
+
+    pub(super) fn iter(&self) -> impl Iterator<Item = (InstrId, &V)> {
+        self.instructions
+            .iter()
+            .enumerate()
+            .map(|(idx, value)| (InstrId(idx), value))
+    }
+
+    pub(super) fn iter_mut(&mut self) -> impl Iterator<Item = (InstrId, &mut V)> {
+        self.instructions
+            .iter_mut()
+            .enumerate()
+            .map(|(idx, value)| (InstrId(idx), value))
+    }
+
+    pub(super) fn get_first_mut(&mut self) -> (InstrId, &mut V) {
+        let idx = 0;
+        (
+            InstrId(idx),
+            self.instructions
+                .get_mut(idx)
+                .expect("Getting first instruction failed: No instructions in map."),
+        )
+    }
+
+    pub(super) fn get_last_mut(&mut self) -> (InstrId, &mut V) {
+        let idx = self.instructions.len() - 1;
+        (
+            InstrId(idx),
+            self.instructions
+                .get_mut(idx)
+                .expect("Getting last instruction failed: No instructions in map."),
+        )
+    }
+
+    pub(super) fn map<F, U>(self, f: F) -> InstrMap<U>
+    where
+        F: Fn(&V) -> U,
+    {
+        let length = self.instructions.len();
+        let mut new_map = InstrMap {
+            instructions: Vec::with_capacity(length),
+        };
+
+        for (_, instr) in self.iter() {
+            new_map.instructions.push(f(instr));
+        }
+
+        new_map
+    }
+}
+
+pub(super) struct InstrMapBuilder {
+    capacity: usize,
+    instructions: Vec<LoweredInstruction>,
+}
+
+impl InstrMapBuilder {
+    pub(super) fn new_with_capacity(capacity: usize) -> Self {
```

Comment:
Nit: The constructor function is commonly called `with_capacity`. See in [std](https://doc.rust-lang.org/stable/std/?search=with_capacity).


```suggestion
    pub(super) fn with_capacity(capacity: usize) -> Self {
```

Reply:
lost diff but done